src/apple/xcode-manager.ts

```diff
@@ -302,103 +331,971 @@ export class XcodeProject {
+   *
+   * @param groupId - The ID of the group to check
+   * @returns True if the group is the main group, false otherwise
+   */
+  private isMainGroup(groupId: string): boolean {
+    return Object.values(this.objects.PBXProject ?? {}).some((project) => {
+      if (typeof project !== 'object') {
+        return false;
+      }
+      return project.mainGroup === groupId;
+    });
+  }
+
+  private getExceptionSetsForGroup(
+    group: XcodeProjectObjectWithId<PBXFileSystemSynchronizedRootGroup>,
+  ): XcodeProjectObjectWithId<PBXFileSystemSynchronizedBuildFileExceptionSet>[] {
+    const exceptions = group.obj.exceptions ?? [];
+    const exceptionSets: XcodeProjectObjectWithId<PBXFileSystemSynchronizedBuildFileExceptionSet>[] =
+      [];
+    for (const exception of exceptions) {
+      const exceptionSet =
+        this.objects.PBXFileSystemSynchronizedBuildFileExceptionSet?.[
+          exception.value
+        ];
+      if (typeof exceptionSet !== 'object') {
+        continue;
+      }
+      exceptionSets.push({
+        id: exception.value,
+        obj: exceptionSet,
+        comment: exception.comment,
+      });
+    }
+    return exceptionSets;
+  }
+
+  // ================================ SYSTEM HELPERS ================================
+
+  private buildProductsDir: string | undefined;
+  /**
+   * Returns the path to the build products directory for the project.
+   *
+   * @returns The path to the build products directory for the project, or undefined if the path is not found
+   */
+  private getBuildProductsDirectoryPath(): string | undefined {
+    if (this.buildProductsDir) {
+      return this.buildProductsDir;
+    }
+    try {
+      const output = execSync(
+        `xcodebuild -project "${this.xcodeprojPath}" -showBuildSettings`,
+        {
+          encoding: 'utf8',
+        },
+      ).trim();
+      // --- Example Output: ---
+      // Command line invocation:
+      // /Applications/Xcode-16.3.0.app/Contents/Developer/usr/bin/xcodebuild -project ./fixtures/test-applications/apple/project-with-synchronized-folders/Project.xcodeproj -showBuildSettings
+      //
+      // Build settings for action build and target Project:
+      //     ACTION = build
+      //     ALLOW_BUILD_REQUEST_OVERRIDES = NO
+      //     ALLOW_TARGET_PLATFORM_SPECIALIZATION = YES
+      //     ALTERNATE_GROUP = staff
+      //     ...
+      const lines = output.split('\n');
+      const settings: Record<string, string> = {};
+      lines.forEach((line) => {
+        const match = line.match(/^\s*(\w+)\s+=\s+(.*)$/);
+        if (match) {
+          settings[match[1]] = match[2];
+        }
+      });
+      this.buildProductsDir =
+        settings['CONFIGURATION_BUILD_DIR'] ??
+        settings['TARGET_BUILD_DIR'] ??
+        settings['BUILD_DIR'];
+
+      return this.buildProductsDir;
+    } catch (error) {
+      debug(`Failed to find build products directory path: ${error as string}`);
+      return undefined;
+    }
+  }
```

Comment:
`m`: This looks similar to the code in `macos-system-helper.ts`. Shouldn't we also move it there?

Reply:
Yes it makes sense to move the `xcodebuild` execution there too. Thanks for pointing this out.